environments/centos7-base/main.py

```diff
@@ -0,0 +1,204 @@
+# write it to disk, send everything to S3, and exit the program.
+#
+# If succeeded is false, we'll assume things failed badly enough that we shouldn't
+# even try to read /grade/results/results.json.
+def finish(succeeded, info):
+    if info['job_id'] is None:
+        # Someone screwed up bad; without a job ID, we can't upload archives to
+        # S3 or notify the webhook that we've failed
+        error('job_id was not specified, so it\'s impossible to upload an archive')
+        sys.exit(1)
+
+    data = None
+    if succeeded and os.path.isfile('/grade/results/results.json'):
+        # Let's try to read their results file
+        # They successfully wrote out the results file!
+        with open('/grade/results/results.json') as json_data:
+            try:
+                data = json.load(json_data)
+            except JSONDecodeError:
+                data = None
+
+    results = {}
+    results['data'] = data
+    results['start_time'] = info['start_time']
+    results['end_time'] = datetime.utcnow().isoformat()
+
+    # Note the distinction between grade/results/results.json and grade/results.json
+    # The former is generated by the code that we're calling and will store
+    # test results; the latter will hold the results file that we'll generate
+    # that contains both test results and some addition metadata
+    with open('grade/results.json', 'w+') as results_file:
+        json.dump(results, results_file)
+
+    # Now we can upload this results file to AWS
+    if info['results_bucket']:
+        s3_results_file = Template('s3://$bucket/job_$job.json').substitute(bucket=info['results_bucket'], job=info['job_id'])
+        s3_results_push_ret = call(['aws', 's3', 'cp', '/grade/results.json', s3_results_file])
+        if s3_results_push_ret is not 0:
+            error('could not push results to S3')
+    else:
+        error('the results bucket was not specified')
+
+    # Now let's make an archive of the /grade directory and send it back to S3
+    # for storage
+    if info['archives_bucket']:
+        zip_ret = call(['tar', '-zcf', '/archive.tar.gz', '/grade/'])
+        if zip_ret is not 0:
+            error('error zipping up archive')
+        else:
+            s3_archive_file = Template('s3://$bucket/job_$job.tar.gz').substitute(bucket=info['archives_bucket'], job=info['job_id'])
+            s3_archive_push_ret = call(['aws', 's3', 'cp', '/archive.tar.gz', s3_archive_file])
+            if s3_archive_push_ret is not 0:
+                error('could not push archive to S3')
+    else:
+        error('the archives bucket was not specified')
+
+    # Finally, notify the webhook (if specified) that this grading job is done
+    if info['webhook_url']:
+        req = urllib2.Request(webhook_url)
+        req.add_header('Content-Type', 'application/json')
+
+        # response = urllib2.urlopen(req, )
+
+    # We're all done now.
+    sys.exit(0 if succeeded else 1)
```

Comment:
Normally we don't explicitly call `exit(0)` on success but rather just let the code end. Calling `exit(1)` on error is correct. In this case by calling `exit(0)` here on success we skip the `log('finishing')` call in `main()`, which probably isn't intended.